File: services/api/src/aec_api/routers/prefab.py
"""R23-PREFAB-KIT — the prefabrication-kit register (project-scoped).

Joins the `prefab_kit` records to the model index (scope + BOM), the pull-plan task they install
under, and the delivery that has to land first. `POST .../freeze` is the one write: it records the
GlobalId list a released kit is fabricated against.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from .. import model_index, rbac
from .. import modules as me
from .. import prefab_kit as pk
from ..db import get_db
from ..rbac import require_role

router = APIRouter()


def _index(pid: str):
    """The project's property index, or None. None is a first-class answer here — a kit register with
    no model still lists its kits and says that nothing about their scope can be checked, which is
    more useful than a 409 that hides the records."""
    try:
        model_index.ensure_loaded(pid)
    except Exception:                       # noqa: BLE001 — no model uploaded is not an error here
        return None
    return model_index.get_index(pid)


def _by_id(db: Session, key: str, pid: str) -> dict[str, dict]:
    if key not in me.TABLES:
        return {}
    return {str(r.get("id")): r for r in me.list_records(db, key, pid, limit=100000)}


@router.get("/projects/{pid}/prefab/kits")
def prefab_register(pid: str, as_of: str | None = None, limit: int = Query(500, ge=1, le=2000),
                    db: Session = Depends(get_db), _: str = Depends(require_role("viewer"))):
    """Every prefab kit on the project, worst first, each with its scope, BOM, scope drift and the
    specific things blocking it.

    Kits sort by the severity of their worst blocker rather than by date: a kit whose released scope
    has **drifted** from what the selector now matches outranks one that is merely late, because a
    late kit is a known problem and a drifted one is an unknown wrong one."""
    if "prefab_kit" not in me.TABLES:
        raise HTTPException(404, "the prefab_kit module is not installed")
    kits = me.list_records(db, "prefab_kit", pid, limit=limit)
    return pk.register(kits, _index(pid),
                       tasks_by_id=_by_id(db, "pull_plan_task", pid),
                       deliveries_by_id=_by_id(db, "delivery", pid),
                       today=as_of)


@router.get("/projects/{pid}/prefab/kits/{rid}")
def prefab_kit_detail(pid: str, rid: str, as_of: str | None = None,
                      db: Session = Depends(get_db), _: str = Depends(require_role("viewer"))):
    """One kit, fully resolved."""
    kit = me.get_record(db, "prefab_kit", pid, rid)
    d = kit.get("data") or {}
    tasks = _by_id(db, "pull_plan_task", pid)
    dels = _by_id(db, "delivery", pid)
    return pk.assess(kit, _index(pid),
                     pull_task=tasks.get(str(d.get("pull_task") or "")),
                     delivery=dels.get(str(d.get("delivery") or "")),
                     today=as_of)


@router.post("/projects/{pid}/prefab/kits/{rid}/freeze")
def prefab_freeze(pid: str, rid: str, db: Session = Depends(get_db),
                  actor: str = Depends(require_role("editor"))):
    """Freeze the kit's scope: resolve the selector once and write the GlobalId list onto the record.

    This is what makes a released kit a document rather than a query. From here the shop builds the
    frozen list; the selector is only the record of how that list was arrived at, and any later
    divergence between the two is reported by the register instead of silently changing what is being
    fabricated.

    Refuses to freeze an empty, erroring or truncated result — a released kit with nothing frozen is
    precisely the failure this whole mechanism exists to prevent, and it would be indistinguishable
    from a correctly released one on every screen."""
    kit = me.get_record(db, "prefab_kit", pid, rid)
    idx = _index(pid)
    # Ask WHY it cannot be frozen and answer with that constant, rather than calling `freeze()` and
    # stringifying whatever it raised. `str(e)` on a caught exception is the py/stack-trace-exposure
    # sink; every string that can reach the response below is a module-level literal.
    blocker = pk.freeze_blocker(kit, idx)
    if blocker:
        raise HTTPException(422, blocker)
    try:
        frozen = pk.freeze(kit, idx)
    except ValueError:                       # deliberately unbound — a fixed literal, never `e`
        raise HTTPException(422, pk.FREEZE_FAILED) from None
    me.update_record(db, "prefab_kit", pid, rid,
                     {"frozen_guids": frozen["frozen_guids"]}, actor,
                     rbac.party_role_for(db, pid, actor))
    return {"ref": kit.get("ref"), "matched": frozen["matched"],
            "selector": frozen["selector"], "frozen": len(frozen["guids"]),
            "note": "the kit now has a written scope; releasing it hands the shop THIS list"}

File: apps/web/src/api/prefab.ts
/** R23-PREFAB-KIT — the prefabrication-kit register, and the one write that makes a kit a document.
 *
 *  Added 2026-09-11 because **all three routes had no client caller at all**. The server side has
 *  been complete since R23: selector resolution against the model index, a BOM, scope-drift
 *  detection, and an eleven-code blocker severity order. None of it was reachable.
 *
 *  Two of the three could not even be *reported* dark: their leaf is `kits`, five characters, and
 *  `services/api/test_route_reachability.py` does not assess a leaf shorter than `MIN_SEGMENT`. The
 *  third, `freeze`, only became visible when that gate stopped letting English prose vouch for a
 *  route — *"…out of the wet/freeze season"* was the sentence standing in for it.
 */
import { HttpCore } from "./httpCore";

type Ctor<T> = new (...args: any[]) => T;

/** One line of a kit's bill of materials: a type, how many, and the quantities that rolled up. */
export interface PrefabBomLine {
  key: string;
  ifc_class: string | null;
  type_name: string | null;
  count: number;
  quantities: Record<string, number>;
}

/** What is stopping this kit. `code` orders the register; `detail` is what a person acts on. */
export interface PrefabBlocker { code: string; detail: string }

/** One kit, resolved against the model. */
export interface PrefabKit {
  id: string | number | null;
  ref: string | null;
  name: string | null;
  trade: string | null;
  fabricator: string | null;
  /** Workflow state. Which list IS the kit depends on it — see `scope_source`. */
  state: string;
  /**
   * `"selector"` before release, `"frozen"` after. This is not cosmetic: a released kit whose
   * scope still comes from the selector is a live query the shop is building against.
   */
  scope_source: "selector" | "frozen";
  selector: { selector: string; guids: string[]; matched: number; truncated?: boolean; error?: string };
  frozen_count: number;
  elements: number;
  bom: { lines: PrefabBomLine[]; totals: Record<string, number>; unresolved: string[] };
  /** Frozen scope vs what the selector matches NOW. `null` until the kit is frozen. */
  drift: { drifted: boolean; added: string[]; removed: string[]; means: string } | null;
  blockers: PrefabBlocker[];
  ready: boolean;
  as_of: string;
  needed_on_site?: string | null;
  pull_task?: { id: string | number | null; ref: string | null; state: string | null;
                open_constraints?: number } | null;
  delivery?: { id: string | number | null; ref: string | null; promised?: string | null } | null;
}

/** The register: every kit, worst first. */
export interface PrefabRegister {
  kits: PrefabKit[];
  total: number;
  ready: number;
  blocked: number;
  blocker_counts: Record<string, number>;
  as_of: string;
  /** False when no model is loaded — every scope figure below is then unverifiable, not zero. */
  model_loaded: boolean;
}

/** What freezing wrote. `frozen` is the count the shop will build against. */
export interface PrefabFreezeResult {
  ref: string | null;
  matched: number;
  selector: string;
  frozen: number;
  note: string;
}

export function withPrefab<TBase extends Ctor<HttpCore>>(Base: TBase) {
  return class extends Base {
    /** Every prefab kit on the project, worst blocker first. */
    prefabRegister(pid: string, asOf?: string) {
      const q = asOf ? `?as_of=${encodeURIComponent(asOf)}` : "";
      return this.json<PrefabRegister>(`/projects/${encodeURIComponent(pid)}/prefab/kits${q}`);
    }
    /** One kit, fully resolved: scope, BOM, drift, installing task and blockers. */
    prefabKit(pid: string, rid: string, asOf?: string) {
      const q = asOf ? `?as_of=${encodeURIComponent(asOf)}` : "";
      return this.json<PrefabKit>(
        `/projects/${encodeURIComponent(pid)}/prefab/kits/${encodeURIComponent(rid)}${q}`);
    }
    /**
     * Freeze the kit's scope: resolve the selector once and write the GlobalId list onto the record.
     *
     * The server refuses an empty, erroring or truncated result with a 422 carrying a fixed literal
     * — a released kit with nothing frozen is indistinguishable from a correct one on every screen,
     * which is the failure the whole mechanism exists to prevent.
     */
    prefabFreeze(pid: string, rid: string) {
      return this.json<PrefabFreezeResult>(
        `/projects/${encodeURIComponent(pid)}/prefab/kits/${encodeURIComponent(rid)}/freeze`,
        { method: "POST" });
    }
  };
}
